src/wooting-usb.c

```diff
@@ -213,192 +228,179 @@ bool wooting_usb_find_keyboard() {
+  // page
+  struct hid_device_info* hid_info_walker = hid_info;
   while (hid_info_walker) {
+    if (connected_keyboards == WOOTING_MAX_RGB_DEVICES) break;
 #ifdef DEBUG_LOG
     printf("Found interface No: %d\n", hid_info_walker->interface_number);
     printf("Found usage page: %d\n", hid_info_walker->usage_page);
 #endif
-    if (hid_info_walker->interface_number == interfaceNr) {
+    if (hid_info_walker->usage_page == CFG_USAGE_PAGE) {
 #ifdef DEBUG_LOG
       printf("Attempting to open\n");
 #endif
       keyboard_handle = hid_open_path(hid_info_walker->path);
       if (keyboard_handle) {
 #ifdef DEBUG_LOG
         printf("Found keyboard_handle: %s\n", hid_info_walker->path);
+        printf("Opened handle: %p\n", keyboard_handle);
 #endif
-// Once the keyboard is found send an init command and abuse two reads to make a
-// 50 ms delay to make sure the keyboard is ready
-#ifdef DEBUG_LOG
-        bool result =
-#endif
-            wooting_usb_send_feature(WOOTING_COLOR_INIT_COMMAND, 0, 0, 0, 0);
-#ifdef DEBUG_LOG
-        printf("Color init result: %d\n", result);
-#endif
-        unsigned char stub = 0;
-        hid_read(keyboard_handle, &stub, 0);
-        hid_read_timeout(keyboard_handle, &stub, 0, 50);
 
-        keyboard_found = true;
-      } else {
-#ifdef DEBUG_LOG
-        printf("No Keyboard handle: %S\n", hid_error(NULL));
-#endif
-      }
-      break;
-    }
-
-    hid_info_walker = hid_info_walker->next;
-  }
-#else
+        // Update pointer array and meta
+        keyboard_handle_array[connected_keyboards] = keyboard_handle;
+        meta_func();
+        wooting_usb_meta->connected = true;
 
-  // We can just search for the interface with matching custom Wooting Cfg usage
-  // page
-  struct hid_device_info *hid_info_walker = hid_info;
-  while (hid_info_walker) {
-#ifdef DEBUG_LOG
-    printf("Found interface No: %d\n", hid_info_walker->interface_number);
-    printf("Found usage page: %d\n", hid_info_walker->usage_page);
-#endif
-    if (hid_info_walker->usage_page == CFG_USAGE_PAGE) {
+        // Any feature sends need to be done after the meta is set so the correct value
+        // for v2_interface is set
 
+        // Once the keyboard is found send an init command
 #ifdef DEBUG_LOG
-      printf("Attempting to open\n");
+        bool result =
 #endif
-      keyboard_handle = hid_open_path(hid_info_walker->path);
-      if (keyboard_handle) {
+          wooting_usb_send_feature(WOOTING_COLOR_INIT_COMMAND, 0, 0, 0, 0);
 #ifdef DEBUG_LOG
-        printf("Found keyboard_handle: %s\n", hid_info_walker->path);
+        printf("Color init result: %d\n", result);
 #endif
 
-        keyboard_found = true;
-      } else {
+        wooting_usb_meta->layout = wooting_usb_get_layout();
+
+        // Increment found keyboard count and switch to the next element in the array
+        connected_keyboards++;
+        wooting_usb_select_device(connected_keyboards);
```

Comment:
How will the layout value of wooting_usb_meta be valid for a particular device when the `wooting_usb_meta` pointer only gets updated with the reference to this device's object after it's set the layout?

Reply:
because the layout is the last thing set before we switch to the next device pointer

Reply:
ah, okay. With passing the pointer to the meta functions this can be cleaned up a bit to make it more resilient to future changes